rust/worker/src/segment/types.rs

```diff
@@ -1,29 +1,362 @@
+            Some(doc) => Some(doc.as_str()),
+            None => None,
+        };
+
+        let embedding = match &log_record.embedding {
+            Some(embedding) => Some(embedding.as_slice()),
+            None => {
+                return Err(LogMaterializerError::EmbeddingMaterializationError);
+            }
+        };
+
+        Ok(Self {
+            data_record: None,
+            offset_id,
+            user_id: Some(user_id),
+            final_operation: Operation::Add,
+            metadata_to_be_merged: metadata,
+            final_document: document,
+            final_embedding: embedding,
+        })
+    }
+}
+
+pub(crate) struct LogMaterializer<'a> {
+    // Is None when record segment is uninitialized.
+    record_segment_reader: Option<RecordSegmentReader<'a>>,
+    logs: Chunk<LogRecord>,
+    curr_max_offset_id: Arc<AtomicU32>,
+}
+
+impl<'a> LogMaterializer<'a> {
     pub(crate) fn new(
-        segment_offset_id: u32,
-        log_record: &'a LogRecord,
-        materialized_record: DataRecord<'a>,
+        record_segment_reader: Option<RecordSegmentReader<'a>>,
+        logs: Chunk<LogRecord>,
+        curr_max_offset_id: Arc<AtomicU32>,
     ) -> Self {
         Self {
-            segment_offset_id,
-            log_record,
-            materialized_record,
+            record_segment_reader,
+            logs,
+            curr_max_offset_id,
         }
     }
+    pub(crate) async fn materialize(
+        &'a self,
+    ) -> Result<Chunk<MaterializedLogRecord<'a>>, LogMaterializerError> {
+        // Populate entries that are present in the record segment.
+        let mut existing_id_to_materialized: HashMap<&str, MaterializedLogRecord> = HashMap::new();
+        let mut new_id_to_materialized: HashMap<&str, MaterializedLogRecord> = HashMap::new();
+        // If record segment is uninitialized then there's nothing in the record segment
+        // yet.
+        if self.record_segment_reader.is_some() {
+            for (log_record, _) in self.logs.iter() {
+                let mut exists: bool = false;
+                match self
+                    .record_segment_reader
+                    .as_ref()
+                    .unwrap()
+                    .data_exists_for_user_id(log_record.record.id.as_str())
+                    .await
+                {
+                    Ok(res) => exists = res,
+                    Err(e) => {
+                        return Err(LogMaterializerError::RecordSegmentError(e));
+                    }
+                };
+                if exists {
+                    match self
+                        .record_segment_reader
+                        .as_ref()
+                        .unwrap()
+                        .get_data_and_offset_id_for_user_id(log_record.record.id.as_str())
+                        .await
+                    {
+                        Ok((data_record, offset_id)) => {
+                            existing_id_to_materialized.insert(
```

Comment:
one that concerns me about this is that we are reordering the log. this doesn't matter from a correctness perspective, but can make certain order-dependent things difficult to debug / reproduce (such as inserts into hnsw)